test/KeyConnector.Tests/Services/Pkcs11RsaKeyServiceTests.cs

```diff
@@ -0,0 +1,571 @@
+            {
+                Pkcs11Provider = "yubihsm", Pkcs11SlotTokenSerialNumber = "chosenSerialNumber",
+            };
+            interopFactory.LoadPkcs11Library(default, default).ReturnsForAnyArgs(library);
+        });
+
+        // Act
+        try
+        {
+            await sut.DecryptAsync(_data);
+        }
+        catch
+        {
+        }
+
+        // Assert
+        slot1.Received().OpenSession(Arg.Any<SessionType>());
+    }
+
+    [Theory]
+    [InlineData(null, CKU.CKU_USER)]
+    [InlineData("so", CKU.CKU_SO)]
+    [InlineData("context_specific", CKU.CKU_CONTEXT_SPECIFIC)]
+    public async Task DecryptAsync_UsesCorrectUserType_GivenInSettings(string userTypeSetting, CKU expectedUserType)
+    {
+        // Create mocks
+        var library = Substitute.For<IPkcs11Library>();
+        var slot = library.AddSlot("chosenSerialNumber");
+        var session = slot.AddSession();
+
+        // Initialize sut
+        var sut = InitializeService((_, _, interopFactory, settings) =>
+        {
+            settings.RsaKey = new KeyConnectorSettings.RsaKeySettings
+            {
+                Pkcs11Provider = "yubihsm",
+                Pkcs11SlotTokenSerialNumber = "chosenSerialNumber",
+                Pkcs11LoginUserType = userTypeSetting,
+                Pkcs11LoginPin = "loginPin"
+            };
+            interopFactory.LoadPkcs11Library(default, default).ReturnsForAnyArgs(library);
+        });
+
+        // Act
+        try
+        {
+            await sut.DecryptAsync(_data);
+        }
+        catch
+        {
+        }
+
+        // Assert
+        session.Received().Login(expectedUserType, "loginPin");
+    }
+
+    [Theory]
+    [InlineData(123UL, CKA.CKA_ID)]
+    [InlineData(null, CKA.CKA_LABEL)]
+    public async Task DecryptAsync_UsesPrivateKeyId_IfAvailableInSettings(ulong? id, CKA expectedType)
+    {
+        // Create mocks
+        var library = Substitute.For<IPkcs11Library>();
+        var slot = library.AddSlot("chosenSerialNumber");
+        var session = slot.AddSession();
+        session.AddPrivateKey();
+
+        IPkcs11InteropFactory pkcs11InteropFactory = null;
+
+        // Initialize sut
+        var sut = InitializeService((_, _, interopFactory, settings) =>
+        {
+            settings.RsaKey = new KeyConnectorSettings.RsaKeySettings
+            {
+                Pkcs11Provider = "yubihsm",
+                Pkcs11SlotTokenSerialNumber = "chosenSerialNumber",
+                Pkcs11LoginUserType = "userType",
+                Pkcs11LoginPin = "loginPin",
+                Pkcs11PrivateKeyId = id,
+                Pkcs11PrivateKeyLabel = "label"
+            };
+            interopFactory.LoadPkcs11Library(default, default).ReturnsForAnyArgs(library);
+            pkcs11InteropFactory = interopFactory;
+        });
+
+        // Act
+        await sut.DecryptAsync(_data);
+
+        // Assert
+        if (id is not null)
```

Comment:
⛏️ : I try to avoid using `if` in the assert section. I think it reduces readability since from the if statement I don't know which of the `inline` tests is for which assert. I think splitting them into separate tests is better. 

The test on ln 490 also uses an `if`, same general sentiment.